x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -0,0 +1,263 @@
+     * average token frequency since we don't have a unique inter-segment token count.
+     * Once we have the maximum number of unique tokens, we use the total count of tokens in the index to calculate
+     * the average frequency ratio.
+     *
+     * @param reader
+     * @param fieldDocCount
+     * @return float
+     * @throws IOException
+     */
+    private float getAverageTokenFreqRatio(IndexReader reader, int fieldDocCount) throws IOException {
+        int numUniqueTokens = 0;
+        for (var leaf : reader.getContext().leaves()) {
+            var terms = leaf.reader().terms(fieldName);
+            if (terms != null) {
+                numUniqueTokens = (int) Math.max(terms.size(), numUniqueTokens);
+            }
+        }
+        if (numUniqueTokens == 0) {
+            return 0;
+        }
+        // Return the sum of TermsEnum.docFreq()
+        return (float) reader.getSumDocFreq(fieldName) / fieldDocCount / numUniqueTokens;
+    }
+
+    /**
+     * Returns true if the token should be queried based on the {@code ratioThreshold} and {@code weightThreshold}
+     * set on the query.
+     */
+    private boolean shouldKeepToken(
+        IndexReader reader,
+        WeightedToken token,
+        int fieldDocCount,
+        float averageTokenFreqRatio,
+        float bestWeight
+    ) throws IOException {
+        if (threshold == null) {
+            return true;
+        }
+        int docFreq = reader.docFreq(new Term(fieldName, token.token()));
+        if (docFreq == 0) {
+            return false;
+        }
+        float tokenFreqRatio = (float) docFreq / fieldDocCount;
+        return tokenFreqRatio < threshold.getRatioThreshold() * averageTokenFreqRatio
+            || token.weight() > threshold.getWeightThreshold() * bestWeight;
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        final MappedFieldType ft = context.getFieldType(fieldName);
+        if (ft == null) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
+        }
+        var qb = new BooleanQuery.Builder();
+        int fieldDocCount = context.getIndexReader().getDocCount(fieldName);
+        float bestWeight = 0f;
+        for (var t : tokens) {
+            bestWeight = Math.max(t.weight(), bestWeight);
+        }
+        float averageTokenFreqRatio = getAverageTokenFreqRatio(context.getIndexReader(), fieldDocCount);
+        if (averageTokenFreqRatio == 0) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against an empty field");
+        }
+        for (var token : tokens) {
+            boolean keep = shouldKeepToken(context.getIndexReader(), token, fieldDocCount, averageTokenFreqRatio, bestWeight);
+            if (threshold != null) {
+                keep ^= threshold.isOnlyScorePrunedTokens();
```

Comment:
For my understanding: in the use case where we'd discard the token for some reason, but `isOnlyScorePrunedTokens()` is true, we keep it instead?

Reply:
Right - 

If `isOnlyScorePrunedTokens()` returns false (the default), then we only keep tokens if `shouldKeepToken` returns true. 
If `isOnlyScorePrunedTokens()` returns true the inverse is true and we only keep tokens that we would have pruned/discarded. This should only ever be used during the rescore phase.